v2/features/farms/FarmsTableRowBody.tsx

```diff
@@ -5,47 +5,73 @@ import { InformationCircleIcon } from "@heroicons/react/solid";
 
 import { useAppSelector } from "v2/store";
 import Link from "v2/components/Link";
-import { CoreSelectors, JarWithData } from "v2/store/core";
+import { BrineryWithData, CoreSelectors, JarWithData } from "v2/store/core";
 import { UserSelectors } from "v2/store/user";
 import { getUserAssetDataWithPrices } from "v2/utils/user";
 import FarmsTableRowDetails from "./FarmsTableRowDetails";
 import FarmsTableRowBodyTransactionControls from "./FarmTableRowBodyTransactionControls";
 import ConnectButton from "./ConnectButton";
 import { useAccount, useNeedsNetworkSwitch } from "v2/hooks";
 import FarmsTableRowBodyV3TransactionControls from "./FarmTableRowBodyTransactionControlsUniV3";
+import BrineryTableRowBodyTransactionControls from "../brinery/BrineryTableRowBodyTransactionControls";
+import BrineryTableRowDetails from "../brinery/BrineryTableRowDetails";
 
 interface Props {
-  jar: JarWithData;
+  jarOrBrinery: JarWithData | BrineryWithData;
   hideDescription?: boolean;
 }
 
-const FarmsTableRowBody: FC<Props> = ({ jar, hideDescription }) => {
+const FarmsTableRowBody: FC<Props> = ({ jarOrBrinery, hideDescription }) => {
   const { t } = useTranslation("common");
-  const account = useAccount();
-  const pfcore = useAppSelector(CoreSelectors.selectCore);
-  const userModel = useAppSelector((state) => UserSelectors.selectData(state, account));
-  const { network, needsNetworkSwitch } = useNeedsNetworkSwitch(jar.chain);
+  const { network, needsNetworkSwitch } = useNeedsNetworkSwitch(jarOrBrinery.chain);
 
-  const data = getUserAssetDataWithPrices(jar, pfcore, userModel);
-  const tokensInWallet = data.depositTokensInWallet.tokensVisible;
+  const tokensInWallet = jarOrBrinery.depositTokensInWallet.tokensVisible;
   const depositTokenCountString = t("v2.farms.tokens", { amount: tokensInWallet });
 
-  const analyticsUrl: string | undefined = jar.details?.apiKey
-    ? "/v2/stats/jar?jar=" + jar.details.apiKey
+  const analyticsUrl: string | undefined = jarOrBrinery.details?.apiKey
+    ? "/v2/stats/jar?jar=" + jarOrBrinery.details.apiKey
     : undefined;
 
-  const isUniV3 = jar.protocol === AssetProtocol.UNISWAP_V3;
-  const token0Name = jar.depositToken.components?.[0];
-  const token1Name = jar.depositToken.components?.[1];
-  const userBalanceToken0 = data.walletComponentTokens[token0Name || ""]?.tokensVisible;
-  const userBalanceToken1 = data.walletComponentTokens[token1Name || ""]?.tokensVisible;
+  const isBrinery = !(jarOrBrinery as JarWithData).farm;
```

Comment:
@larrythecucumber321 there's one more trick you could use (or I do it later as a part of my current refactor). I defined a custom type guard in `core.helpers.ts`:

```typescript
export const isBrinery = (asset: Asset | undefined): asset is BrineryWithData =>
  asset?.type === "brinery";
```

Then we can use this to determine if the asset is a brinery anywhere. The benefit of that is also you won't have to cast types to `BrineryWithData`, it will just work™ . It will also make passing around the `isBrinery` prop unnecessary.